lib/src/model/result.dart

```diff
@@ -12,8 +12,8 @@ class Result<Value, Failure> {
         _value = null; // ignore: avoid_field_initializers_in_const_classes
 
   /// Provides the value in case of success or throws [Exception] in case of failure.
-  Value dematerialize() => iif(
-        success: (value) => value!,
+  Value? dematerialize() => iif(
```

Comment:
No way this is right. The type should be non-optional.

Please revert.

Reply:
I know this is crap but can you explain me how we should deal with `Result<void, GenericFailure<Faliure>>` ? Because this returns an object of `Result(value: null, failure: null)` and thus crash. Maybe we should use Optional.none for this

Reply:
This is sort of a language design issue. The fact that the type `void` has the only possible instance, which is `null`, is self-contradicting: you can have a non-optional `void` type, but it's going to contain `null`!

Despite being named `void`, this type in Dart (as well as in other languages, take C for example) has a structure of a "unit" type, which means a type that only has a single possible instance (hence "unit"). A type with no possible instances should be called "void" (like `Never` in Swift), it is not possible to have an instance of such type (it is still useful for functions that never return [like `fatalError(·)`] and you can tell that from their type, that's really cool).

But that's weird naming with historical reasons.

I hope you see the contradiction in Dart's "sound null-safety" here 🙂

The thing is, here we need a _proper_ "unit" type with a single possible instance, which is _not null_.

We already have such a type, see [lib/src/model/unit.dart](https://github.com/PhilipsHue/flutter_reactive_ble/blob/3e8ce1c85b111d0f89e2109c61df8c8a45829f97/lib/src/model/unit.dart), and it's used in [PluginController.clearGattCache(·)](https://github.com/PhilipsHue/flutter_reactive_ble/blob/fc94bb468766339edff2e250c771fdefc24e729b/lib/src/plugin_controller.dart#L348).

I would suggest to start using the `Unit` more in public types (in public interfaces), because with Dart null-safety it apparently becomes a necessity.

Reply:
damm I was thinking yesterday night if only we had unit 🤦